### DISCUSSION TITLE
Size memory too big

### DISCUSSION THREAD

**Discussion by nvhhung** [ben-manes/caffeine#813] [Q&A] [answered]
I have a issue with caffein. I use AsyncCache: 
```java
var caffeine = Caffeine.newBuilder()
                .maximumSize(maximumSize)
                .expireAfter(new Expiry<String, CachingMsg>() {
                    @Override
                    public long expireAfterCreate(String key, CachingMsg value, long currentTime) {
                        var seconds = (value.getExpireTime() - System.currentTimeMillis()) / 1000L;
                        return TimeUnit.SECONDS.toNanos(Math.max(seconds, 0));
                    }

                    @Override
                    public long expireAfterUpdate(String key, CachingMsg value, long currentTime, @NonNegative long currentDuration) {
                        var seconds = (value.getExpireTime() - System.currentTimeMillis()) / 1000L;
                        return TimeUnit.SECONDS.toNanos(Math.max(seconds, 0));
                    }

                    @Override
                    public long expireAfterRead(String key, CachingMsg value, long currentTime, @NonNegative long currentDuration) {
                        return currentDuration;
                    }
                })
                .recordStats()
                .scheduler(Scheduler.systemScheduler())
                .removalListener((key, value, cause) -> log.info(null,String.format("%s: key %s of mainKey %s", cause, key, mainKey)))
```
 Data i save in cache : "{\"status\":\"1\",\"lastUpdate\":1667883430507}"
 But when i save 2M record, memory usage reach 1.5BG => not optimize.
 I hope team support me resolve this issue, pls!!

**Comment by ben-manes** [comment, marked as answer]
What is the problem? At 2M entries that is 750 bytes per, which seems reasonable. It sounds like you oversized if that is too much

**Reply by nvhhung** [reply]
My problem, after expried time, memory usage no reduce. Still in 1GB, can you suggest solve problem with init cache ??

**Comment by nvhhung** [comment]
I dump memory usage in local : 
- Before expire time have : 
![Screenshot from 2022-11-08 16-32-40](https://user-images.githubusercontent.com/74722963/200528366-f357457c-c2ae-4458-9bbc-5f9c5b08968d.png)
- Affter expire time: 
![Screenshot from 2022-11-08 16-33-29](https://user-images.githubusercontent.com/74722963/200528554-174107ef-c002-4c7e-9175-2a8aa437d3a0.png)

I see class java.util.concurrent.ConcurrentHashMap exist in memory. Whether it task my memory ??. Pls team sp me.

**Reply by ben-manes** [reply]
That says that there are only 298 entries that combined uses 19.8mb of memory. That seems to indicate that the cache did evict the entries and that they were garbage collected.

**Comment by nvhhung** [comment]
I have built chart for my system. 
In 13h53, memory usage increase suddenly. Meanwhile traffic of my server is stable.
![Screenshot from 2022-11-09 14-29-11](https://user-images.githubusercontent.com/74722963/200766289-6591770c-6d73-43f9-978d-765faf316306.png)
![Screenshot from 2022-11-09 14-32-15](https://user-images.githubusercontent.com/74722963/200767059-fff43bc3-081b-4663-afd7-9c536591208f.png)
At 13h52 my entries is : 
![Screenshot from 2022-11-09 14-35-27](https://user-images.githubusercontent.com/74722963/200767313-bf606dae-72ae-46ac-bb03-1f1e1307bee9.png)
But at 13h54 my entries is: 
![Screenshot from 2022-11-09 14-36-37](https://user-images.githubusercontent.com/74722963/200767639-77c3e491-91b7-4842-a70d-cddee0a55b89.png)

Size of entry no significant increase. 

Can you tell me what happens with my system ?

**Comment by nvhhung** [comment]
![Screenshot from 2022-11-09 14-41-43](https://user-images.githubusercontent.com/74722963/200768469-3dde0064-98e4-4e17-a803-3d0c6646ff1c.png)
Chart eviction at this time.

**Comment by ben-manes** [comment]
I cannot help from that information. I don't know your system, the eviction and cache sizes seem stable, and there isn't anything to indicate a change needed to this library. Depending on your JDK version, only the most recent will return memory once allocated so it is unclear if that is why it stays high (heap vs system allocated). Usually it takes a combination of logs, heap dumps, and JFR snapshots to figure these types of application problems.

**Comment by nvhhung** [comment]
Thank you very much. I appreciate you !
I have 2 question, hope you answer. 
Q1: When all entries evitcted i don't put any entry to cache, my heap size decrease bit, used size decrease so much. 
I want to ask why heap size don't decrease so much to  appropriate used size ?. And why used size increase and then decrease again ?.
![Screenshot from 2022-11-11 11-28-33](https://user-images.githubusercontent.com/74722963/201263668-ec5e3d42-1ed1-4a7c-b49c-05642cf0b0dc.png)
P/s : I use JDK 13, use asyncCache have statregy : expireAfterCreate. My data entry: Key: G3PT46tEwk, value: {"counter":10,"expireTime":0,"data":null,"counterValue":10}. Data type of both is String.

Q2: I put entry have value Object, is there any difference i put value is  String (Convert from object) for Size, Perfomance of Garbage Collection ,... ?

**Reply by ben-manes** [reply]
Q1: The used heap size shows that it as a saw-tooth pattern which is desirable, so it is able to reclaim memory after a garbage collection. The total heap size is how much was allocated from the operating system. This traditionally is claimed and never released even after the usage shrinks, so it only grows up to a maximum. In later JDKs this was improved for some garbage collectors so it depends on your collector and jdk version ([JEP-346](https://openjdk.org/jeps/346), [JEP-351](https://openjdk.org/jeps/351), [JEP-387](https://openjdk.org/jeps/387)).

Q2: From the cache's perspective it's the same, as it is just a layer above `ConcurrentHashMap` to enforce a bounding.

**Comment by nvhhung** [comment]
Sorry for bothering you, i wanna ask u new question.
I'm using AsyncCache for my application. I dumped heap, i see have class CompletableFuture run with caffein. Can it be cause increase memory uasage of my application. Because my application have huge throughput, about 300-500 rqs.
AsyncCache vs Cache, 2 type Cache, which perfomance better, memory usages fewwer ??

**Reply by ben-manes** [reply]
AsyncCache is basically `Cache<K, CompletableFuture<V>>`, so it wraps the value. That adds a small amount of overhead per entry. However the performance cost of blocking threads might outweigh that, so which is better is use-case specific.
